crates/bin/pd/src/migrate.rs

```diff
@@ -72,37 +80,96 @@ impl Migration {
         );
         tracing::info!("started migration");
 
-        match self {
-            Migration::ReadyToStart => {
-                reset_halt_bit::migrate(storage, pd_home, genesis_start).await?;
-                return Ok(());
-            }
-            Migration::SimpleMigration => {
-                simple::migrate(storage, pd_home.clone(), genesis_start).await?
-            }
-
-            Migration::Testnet72 => {
-                testnet72::migrate(storage, pd_home.clone(), genesis_start).await?
-            }
-
-            Migration::Testnet74 => {
-                testnet74::migrate(storage, pd_home.clone(), genesis_start).await?
-            }
-
-            Migration::Testnet76 => {
-                testnet76::migrate(storage, pd_home.clone(), genesis_start).await?
-            }
-            Migration::Testnet77 => {
-                testnet77::migrate(storage, pd_home.clone(), genesis_start).await?
-            }
+        // If this is `ReadyToStart`, we need to reset the halt bit and return early.
+        if let Migration::ReadyToStart = self {
+            reset_halt_bit::migrate(storage, pd_home, genesis_start).await?;
+            return Ok(());
+        }
+
+        // Collect various pieces of state pre-migration:
+        let initial_state = storage.latest_snapshot();
+        let root_hash = initial_state
+            .root_hash()
+            .await
+            .expect("chain state has a root hash");
+        let pre_upgrade_root_hash: RootHash = root_hash.into();
+        let pre_upgrade_height = initial_state
+            .get_block_height()
+            .await
+            .expect("chain state has a block height");
+        let post_upgrade_height = pre_upgrade_height.wrapping_add(1);
+
+        let migration_duration = match self {
+            Migration::SimpleMigration => simple::migrate(storage.clone()).await?,
+
+            Migration::Testnet78 => testnet78::migrate(storage.clone()).await?,
+            _ => unreachable!(),
+        };
```

Comment:
I don't think we should do this specific refactor, it doesn't address the module's main issues (brittleness) but it shuffles important code paths around, right after we tested them, but ahead of having integration tests ready (#4337).

In its current state, it makes genesis generation work against a stale snapshot which is a bug. It would be caught by integration tests but this is an example of how deliberate we have to be about reworking this, especially now.

Reply:
I'm mixed on whether to refactor right now or not. I hear you about the testing, however adding the new migration seems to require a careful copy-pasting of genesis handling code and manual testing anyways. I like your suggestion of moving everything into a `Trait` impl.

Reply:
Reverted the refactoring for now